pkg/maps/lbmap/maglev_outer_map.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lbmap
+
+import (
+	"unsafe"
+
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/byteorder"
+	"github.com/cilium/cilium/pkg/ebpf"
+)
+
+const (
+	// UnknownMaglevTableSize is a constant that represents an unknown table
+	// size for a maglev outer map.
+	UnknownMaglevTableSize = 0
+)
+
+// maglevOuterMap is the internal representation of a maglev outer map.
+type maglevOuterMap struct {
+	*ebpf.Map
+	TableSize uint32
+}
+
+// MaglevOuterKey is the key of a maglev outer map.
+type MaglevOuterKey struct {
+	RevNatID uint16
+}
+
+// MaglevOuterVal is the value of a maglev outer map.
+type MaglevOuterVal struct {
+	FD uint32
+}
+
+// NewMaglevOuterMap returns a new object representing a maglev outer map.
+func NewMaglevOuterMap(name string, maxEntries int, tableSize uint32, innerMap *ebpf.MapSpec) (*maglevOuterMap, error) {
+	m := ebpf.NewMap(&ebpf.MapSpec{
+		Name:       name,
+		Type:       ebpf.HashOfMaps,
+		KeySize:    uint32(unsafe.Sizeof(MaglevOuterKey{})),
+		ValueSize:  uint32(unsafe.Sizeof(MaglevOuterVal{})),
+		MaxEntries: uint32(maxEntries),
+		InnerMap:   innerMap,
+		Pinning:    ebpf.PinByName,
+	})
+
+	if err := m.OpenOrCreate(); err != nil {
+		return nil, err
+	}
+
+	return &maglevOuterMap{
+		Map:       m,
+		TableSize: tableSize,
+	}, nil
+}
+
+// MaglevOuterMapTableSize tries to determine the table size of a given maglev
+// outer map.
+//
+// The function returns:
+// - a bool indicating whether the outer map exists or not
+// - an integer indicating the table size. In case the table size cannot be
+//   determined, the UnknownMaglevTableSize constant (0) is returned.
+func MaglevOuterMapTableSize(mapName string) (bool, uint32) {
+	prevMap, err := ebpf.LoadPinnedMap(bpf.MapPath(mapName), nil)
+	if err != nil {
+		// No outer map found.
+		return false, UnknownMaglevTableSize
+	}
+	defer prevMap.Close()
+
+	var firstKey MaglevOuterKey
+	err = prevMap.NextKey(nil, &firstKey)
```

Comment:
nit: as `err` is only used in the following `if` statement, you can use

```go
if err := prevMap.NextKey(nil, &firstKey); err != nil {
    // ...
}
```

Similarly below.